src/precompiles/modexp.rs

```diff
@@ -132,24 +66,96 @@ impl Precompile for ModExp<Byzantium> {
             }
         };
 
-        Ok(PrecompileOutput::without_logs(cost, output))
+        Ok(output)
+    }
+}
+
+impl ModExp<Byzantium> {
+    fn mul_complexity(x: u64) -> Result<u64, ExitError> {
+        if x <= 64 {
+            Ok(x * x)
+        } else if x <= 1_024 {
+            Ok(x * x / 4 + 96 * x - 3_072)
+        } else {
+            let (x_sq, overflow) = x.overflowing_mul(x);
+            if overflow {
+                Err(ExitError::OutOfGas)
+            } else {
+                Ok(x_sq / 16 + 480 * x - 199_680)
+            }
+        }
+    }
+}
+
+impl Precompile for ModExp<Byzantium> {
+    fn required_gas(input: &[u8]) -> Result<u64, ExitError> {
+        let base_len = U256::from(&input[0..32]).as_u64();
+        let exp_len = U256::from(&input[32..64]).as_u64();
+        let mod_len = U256::from(&input[64..96]).as_u64();
+
+        let mul = Self::mul_complexity(core::cmp::max(mod_len, base_len))?;
+        let iter_count = Self::calc_iter_count(exp_len, base_len, &input);
+        let (mut gas, overflow) = mul.overflowing_mul(core::cmp::max(iter_count, 1));
+
+        if overflow {
+            Err(ExitError::OutOfGas)
+        } else {
+            gas /= 20;
+            Ok(gas)
+        }
+    }
+
+    /// See: https://eips.ethereum.org/EIPS/eip-198
+    /// See: https://etherscan.io/address/0000000000000000000000000000000000000005
+    fn run(input: &[u8], target_gas: u64, _context: &Context) -> PrecompileResult {
+        let cost = Self::required_gas(input)?;
+        if cost > target_gas {
+            Err(ExitError::OutOfGas)
+        } else {
+            let output = Self::run_inner(input)?;
+            Ok(PrecompileOutput::without_logs(cost, output))
+        }
+    }
+}
+
+impl ModExp<Berlin> {
+    fn mul_complexity(base_len: u64, mod_len: u64) -> u64 {
+        let max_len = core::cmp::max(mod_len, base_len);
+        let words = max_len.div_ceil(&8);
+        words.pow(2)
```

Comment:
```suggestion
        words.saturating_mul(words)
```

Can use saturating multiplication to avoid overflow.